pkg/pub_worker/lib/src/bin/pana_wrapper.dart

```diff
@@ -197,61 +197,130 @@ Future<(SdkConfig, SdkConfig)> _detectSdks(Pubspec pubspec) async {
       flutterSdks.firstWhereOrNull((sdk) => !sdk.version.isPreRelease) ??
           flutterSdks.firstOrNull;
 
-  // NOTE: this is very simple constraint check right now. Instead, we should:
-  //       - try to use the latest SDKs in the docker image, or
-  //       - try to use the latest downloadable SDKs, or
-  //       - try to use the latest beta/preview SDKs, or
-  //       - fall back to the latest dev/master branch.
-  final needsNewer = needsNewerSdk(
-          sdkVersion: installedDartSdk?.version,
-          constraint: pubspec.dartSdkConstraint) ||
-      needsNewerSdk(
-          sdkVersion: installedFlutterSdk?.version,
-          constraint: pubspec.flutterSdkConstraint);
+  bool matchesSdks({
+    required Version? dart,
+    required Version? flutter,
+    bool allowsMissingVersion = false,
+  }) {
+    if (!allowsMissingVersion && (dart == null || flutter == null)) {
+      return false;
+    }
 
-  var dartSdkPath = installedDartSdk?.path;
-  var flutterSdkPath = installedFlutterSdk?.path;
-  String configKind = 'stable';
-  if (needsNewer) {
-    configKind = 'preview';
-    dartSdkPath = await _previewDartSdk() ??
-        dartSdks.firstWhereOrNull((sdk) => sdk.version.isPreRelease)?.path ??
-        dartSdkPath;
-    flutterSdkPath = await _previewFlutterSdk() ??
-        flutterSdks.firstWhereOrNull((sdk) => sdk.version.isPreRelease)?.path ??
-        flutterSdkPath;
+    return sdkMatchesConstraint(
+          sdkVersion: dart,
+          constraint: pubspec.dartSdkConstraint,
+        ) &&
+        sdkMatchesConstraint(
+          sdkVersion: flutter,
+          constraint: pubspec.flutterSdkConstraint,
+        );
+  }
+
+  // try to use the latest SDKs in the docker image
+  final matchesInstallSdks = matchesSdks(
+    dart: installedDartSdk?.version,
+    flutter: installedFlutterSdk?.version,
+    allowsMissingVersion: true,
+  );
+  if (matchesInstallSdks) {
+    return (
+      SdkConfig(
+        rootPath: installedDartSdk?.path,
+        configHomePath: _configHomePath('dart', 'stable'),
+      ),
+      SdkConfig(
+        rootPath: installedFlutterSdk?.path,
+        configHomePath: _configHomePath('flutter', 'stable'),
+      ),
+    );
   }
 
+  final flutterArchive = await fetchFlutterArchive();
+
+  // try to use the latest stable downloadable SDKs
+  final latestStableDartSdk =
+      await fetchLatestDartSdkVersion(channel: 'stable');
+  final latestStableDartSdkVersion = latestStableDartSdk?.semanticVersion;
+  final matchesLatestStable = matchesSdks(
+    dart: latestStableDartSdkVersion,
+    flutter: flutterArchive?.latestStable?.semanticVersion,
+  );
+  if (matchesLatestStable) {
```

Comment:
This screams for a loop (pseudo-code)
```
var found = false;
for (final sdkVersions in [latestStableVersions, latestBetaVersions]) {
  if (sdkMatchesConstraint(constraint, sdkVersions)) {
    _installSdk(sdkVersions);
    found = true;
    break;
  }
}
if (!found) {
  // use installed
}

Reply:
This is running on each worker and each package-version separately on each run. If we were doing this in a loop, we'd need to query the versions each time (and we don't share a redis cache between the workers), or we'd need to implement this logic outside of the worker where we can cache things. The current implementation does it only on-demand.

Reply:
Not to "run in a loop" just to factor out checking and installing an sdk version, so you can avoid repeating the same code for latest stable and latest beta.

Reply:
Ah, got it, I can give it another go (it got a bit complex on my first try to make it more generic and I ditched it).

Reply:
If it is too complex we can also ship as-is

Reply:
Updated, PTAL.